src/python_testing/TC_DEM_2_7.py

```diff
@@ -0,0 +1,346 @@
+                     "Verify value is 0x00 (NoOptOut)"),
+            TestStep("21", "TH sends RequestConstraintBasedPowerForecast with constraints[0].{StartTime=Forecast.StartTime, Duration=Forecast.Slots[0].DefaultDuration, NominalPower=Forecast.Slots[0].NominalPower, MaximumEnergy=Forecast.Slots[0].NominalEnergy}, Cause=LocalOptimization.",
+                     "Verify DUT responds with status SUCCESS(0x00)"),
+            TestStep("21a", "TH reads Forecast attribute.",
+                     "Value has to include ForecastUpdateReason=LocalOptimization"),
+            TestStep("22", "TH sends CancelRequest.",
+                     "Verify DUT responds with status SUCCESS(0x00)"),
+            TestStep("22a", "TH reads Forecast attribute.",
+                     "Value has to include ForecastUpdateReason=InternalOptimization"),
+            TestStep("23", "TH sends CancelRequest.",
+                     "Verify DUT responds with status INVALID_IN_STATE(0xcb)"),
+            TestStep("24", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for Constraints-based Adjustment Adjustment Test Event Clear.",
+                     "Verify DUT responds with status SUCCESS(0x00)"),
+        ]
+
+        return steps
+
+    @async_test_body
+    async def test_TC_DEM_2_7(self):
+        # pylint: disable=too-many-locals, too-many-statements
+        """Run the test steps."""
+        self.step("1")
+        # Commission DUT - already done
+
+        # Subscribe to Events and when they are sent push them to a queue for checking later
+        events_callback = EventChangeCallback(Clusters.DeviceEnergyManagement)
+        await events_callback.start(self.default_controller,
+                                    self.dut_node_id,
+                                    self.matter_test_config.endpoint)
+
+        self.step("2")
+        await self.check_test_event_triggers_enabled()
+
+        self.step("3")
+        await self.send_test_event_trigger_constraint_based_adjustment()
+
+        self.step("3a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("3b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        logging.info(forecast)
+        asserts.assert_greater(forecast.slots[0].nominalPower, 0)
+        asserts.assert_greater(forecast.slots[0].minPower, 0)
+        asserts.assert_greater(forecast.slots[0].maxPower, 0)
+        asserts.assert_greater(forecast.slots[0].nominalEnergy, 0)
+
+        self.step("3c")
+        await self.check_dem_attribute("OptOutState", Clusters.DeviceEnergyManagement.Enums.OptOutStateEnum.kNoOptOut)
+
+        self.step("4")
+        # Matter UTC is time since 00:00:00 1/1/2000
+        now = int(utc_time_in_matter_epoch()/1000000)
```

Comment:
This should probably either be a const (or whatever those are in Python) or a func that does the conversion.

Reply:
Added a method to the base class TC_DEMTestBase.py to handle this.